{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "97a5d346",
   "metadata": {},
   "source": [
    "# Get analysis data of one study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f0b88dc5-991a-4705-964d-a55ff57d3eb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.pipeline-version</th>\n",
       "      <th>attributes.analysis-status</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.experiment-type</th>\n",
       "      <th>attributes.analysis-summary</th>\n",
       "      <th>attributes.is-private</th>\n",
       "      <th>attributes.complete-time</th>\n",
       "      <th>attributes.instrument-platform</th>\n",
       "      <th>attributes.instrument-model</th>\n",
       "      <th>relationships.study.data.id</th>\n",
       "      <th>relationships.study.data.type</th>\n",
       "      <th>relationships.run.data.id</th>\n",
       "      <th>relationships.run.data.type</th>\n",
       "      <th>relationships.sample.data.id</th>\n",
       "      <th>relationships.sample.data.type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00567538</td>\n",
       "      <td>5.0</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00567538</td>\n",
       "      <td>amplicon</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-09-24T00:39:12</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>MGYS00005611</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRR11282158</td>\n",
       "      <td>runs</td>\n",
       "      <td>SRS6297619</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00567539</td>\n",
       "      <td>5.0</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00567539</td>\n",
       "      <td>amplicon</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-09-24T00:40:20</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>MGYS00005611</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRR11282082</td>\n",
       "      <td>runs</td>\n",
       "      <td>SRS6297694</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00567540</td>\n",
       "      <td>5.0</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00567540</td>\n",
       "      <td>amplicon</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-09-24T00:40:54</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>MGYS00005611</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRR11282054</td>\n",
       "      <td>runs</td>\n",
       "      <td>SRS6297726</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00567541</td>\n",
       "      <td>5.0</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00567541</td>\n",
       "      <td>amplicon</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-09-24T00:41:29</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>MGYS00005611</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRR11282164</td>\n",
       "      <td>runs</td>\n",
       "      <td>SRS6297613</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00567542</td>\n",
       "      <td>5.0</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00567542</td>\n",
       "      <td>amplicon</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-09-24T00:42:01</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>MGYS00005611</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRR11282163</td>\n",
       "      <td>runs</td>\n",
       "      <td>SRS6297614</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00567650</td>\n",
       "      <td>5.0</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00567650</td>\n",
       "      <td>amplicon</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-09-24T01:49:48</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>MGYS00005611</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRR11282052</td>\n",
       "      <td>runs</td>\n",
       "      <td>SRS6297724</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00567651</td>\n",
       "      <td>5.0</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00567651</td>\n",
       "      <td>amplicon</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-09-24T01:50:24</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>MGYS00005611</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRR11282051</td>\n",
       "      <td>runs</td>\n",
       "      <td>SRS6297725</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00567652</td>\n",
       "      <td>5.0</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00567652</td>\n",
       "      <td>amplicon</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-09-24T01:50:59</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>MGYS00005611</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRR11282050</td>\n",
       "      <td>runs</td>\n",
       "      <td>SRS6297727</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00567653</td>\n",
       "      <td>5.0</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00567653</td>\n",
       "      <td>amplicon</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-09-24T01:51:29</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>MGYS00005611</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRR11282049</td>\n",
       "      <td>runs</td>\n",
       "      <td>SRS6297730</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00567654</td>\n",
       "      <td>5.0</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00567654</td>\n",
       "      <td>amplicon</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-09-24T01:52:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>MGYS00005611</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRR11282048</td>\n",
       "      <td>runs</td>\n",
       "      <td>SRS6297728</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              type            id attributes.pipeline-version   \n",
       "0    analysis-jobs  MGYA00567538                         5.0  \\\n",
       "1    analysis-jobs  MGYA00567539                         5.0   \n",
       "2    analysis-jobs  MGYA00567540                         5.0   \n",
       "3    analysis-jobs  MGYA00567541                         5.0   \n",
       "4    analysis-jobs  MGYA00567542                         5.0   \n",
       "..             ...           ...                         ...   \n",
       "112  analysis-jobs  MGYA00567650                         5.0   \n",
       "113  analysis-jobs  MGYA00567651                         5.0   \n",
       "114  analysis-jobs  MGYA00567652                         5.0   \n",
       "115  analysis-jobs  MGYA00567653                         5.0   \n",
       "116  analysis-jobs  MGYA00567654                         5.0   \n",
       "\n",
       "    attributes.analysis-status attributes.accession   \n",
       "0                    completed         MGYA00567538  \\\n",
       "1                    completed         MGYA00567539   \n",
       "2                    completed         MGYA00567540   \n",
       "3                    completed         MGYA00567541   \n",
       "4                    completed         MGYA00567542   \n",
       "..                         ...                  ...   \n",
       "112                  completed         MGYA00567650   \n",
       "113                  completed         MGYA00567651   \n",
       "114                  completed         MGYA00567652   \n",
       "115                  completed         MGYA00567653   \n",
       "116                  completed         MGYA00567654   \n",
       "\n",
       "    attributes.experiment-type   \n",
       "0                     amplicon  \\\n",
       "1                     amplicon   \n",
       "2                     amplicon   \n",
       "3                     amplicon   \n",
       "4                     amplicon   \n",
       "..                         ...   \n",
       "112                   amplicon   \n",
       "113                   amplicon   \n",
       "114                   amplicon   \n",
       "115                   amplicon   \n",
       "116                   amplicon   \n",
       "\n",
       "                           attributes.analysis-summary  attributes.is-private   \n",
       "0    [{'key': 'Submitted nucleotide sequences', 'va...                  False  \\\n",
       "1    [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "2    [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "3    [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "4    [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "..                                                 ...                    ...   \n",
       "112  [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "113  [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "114  [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "115  [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "116  [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "\n",
       "    attributes.complete-time attributes.instrument-platform   \n",
       "0        2020-09-24T00:39:12                       ILLUMINA  \\\n",
       "1        2020-09-24T00:40:20                       ILLUMINA   \n",
       "2        2020-09-24T00:40:54                       ILLUMINA   \n",
       "3        2020-09-24T00:41:29                       ILLUMINA   \n",
       "4        2020-09-24T00:42:01                       ILLUMINA   \n",
       "..                       ...                            ...   \n",
       "112      2020-09-24T01:49:48                       ILLUMINA   \n",
       "113      2020-09-24T01:50:24                       ILLUMINA   \n",
       "114      2020-09-24T01:50:59                       ILLUMINA   \n",
       "115      2020-09-24T01:51:29                       ILLUMINA   \n",
       "116      2020-09-24T01:52:00                       ILLUMINA   \n",
       "\n",
       "    attributes.instrument-model relationships.study.data.id   \n",
       "0              Illumina MiniSeq                MGYS00005611  \\\n",
       "1              Illumina MiniSeq                MGYS00005611   \n",
       "2              Illumina MiniSeq                MGYS00005611   \n",
       "3              Illumina MiniSeq                MGYS00005611   \n",
       "4              Illumina MiniSeq                MGYS00005611   \n",
       "..                          ...                         ...   \n",
       "112            Illumina MiniSeq                MGYS00005611   \n",
       "113            Illumina MiniSeq                MGYS00005611   \n",
       "114            Illumina MiniSeq                MGYS00005611   \n",
       "115            Illumina MiniSeq                MGYS00005611   \n",
       "116            Illumina MiniSeq                MGYS00005611   \n",
       "\n",
       "    relationships.study.data.type relationships.run.data.id   \n",
       "0                         studies               SRR11282158  \\\n",
       "1                         studies               SRR11282082   \n",
       "2                         studies               SRR11282054   \n",
       "3                         studies               SRR11282164   \n",
       "4                         studies               SRR11282163   \n",
       "..                            ...                       ...   \n",
       "112                       studies               SRR11282052   \n",
       "113                       studies               SRR11282051   \n",
       "114                       studies               SRR11282050   \n",
       "115                       studies               SRR11282049   \n",
       "116                       studies               SRR11282048   \n",
       "\n",
       "    relationships.run.data.type relationships.sample.data.id   \n",
       "0                          runs                   SRS6297619  \\\n",
       "1                          runs                   SRS6297694   \n",
       "2                          runs                   SRS6297726   \n",
       "3                          runs                   SRS6297613   \n",
       "4                          runs                   SRS6297614   \n",
       "..                          ...                          ...   \n",
       "112                        runs                   SRS6297724   \n",
       "113                        runs                   SRS6297725   \n",
       "114                        runs                   SRS6297727   \n",
       "115                        runs                   SRS6297730   \n",
       "116                        runs                   SRS6297728   \n",
       "\n",
       "    relationships.sample.data.type  \n",
       "0                          samples  \n",
       "1                          samples  \n",
       "2                          samples  \n",
       "3                          samples  \n",
       "4                          samples  \n",
       "..                             ...  \n",
       "112                        samples  \n",
       "113                        samples  \n",
       "114                        samples  \n",
       "115                        samples  \n",
       "116                        samples  \n",
       "\n",
       "[117 rows x 17 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from jsonapi_client import Session\n",
    "import pandas as pd\n",
    "\n",
    "accession = \"MGYS00005611\"\n",
    "\n",
    "with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    analyses = map(lambda r: r.json, mgnify.iterate(f'studies/{accession}/analyses'))\n",
    "    analyses = pd.json_normalize(analyses)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c486afb3",
   "metadata": {},
   "source": [
    "# Get data of one sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b049fbad-8eac-409d-b332-745b8da2cf56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.sample-metadata</th>\n",
       "      <th>attributes.longitude</th>\n",
       "      <th>attributes.biosample</th>\n",
       "      <th>attributes.latitude</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.analysis-completed</th>\n",
       "      <th>attributes.collection-date</th>\n",
       "      <th>attributes.geo-loc-name</th>\n",
       "      <th>attributes.sample-desc</th>\n",
       "      <th>attributes.environment-biome</th>\n",
       "      <th>attributes.environment-feature</th>\n",
       "      <th>attributes.environment-material</th>\n",
       "      <th>attributes.sample-name</th>\n",
       "      <th>attributes.sample-alias</th>\n",
       "      <th>attributes.last-update</th>\n",
       "      <th>relationships.biome.data.id</th>\n",
       "      <th>relationships.biome.data.type</th>\n",
       "      <th>relationships.studies.data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS2799892</td>\n",
       "      <td>[{'key': 'investigation type', 'value': 'metag...</td>\n",
       "      <td>-76.4716</td>\n",
       "      <td>SAMEA4981828</td>\n",
       "      <td>44.6033</td>\n",
       "      <td>ERS2799892</td>\n",
       "      <td>2019-03-12</td>\n",
       "      <td>2017-01-01</td>\n",
       "      <td>Canada</td>\n",
       "      <td>08-152</td>\n",
       "      <td>Mixedwood_Plains</td>\n",
       "      <td>Envo:00000020</td>\n",
       "      <td>Envo:00002011</td>\n",
       "      <td>08-152</td>\n",
       "      <td>08-152</td>\n",
       "      <td>2023-07-03T10:07:25</td>\n",
       "      <td>root:Environmental:Aquatic:Freshwater:Lake</td>\n",
       "      <td>biomes</td>\n",
       "      <td>[{'id': 'MGYS00003941', 'type': 'studies'}]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type          id                         attributes.sample-metadata   \n",
       "0  samples  ERS2799892  [{'key': 'investigation type', 'value': 'metag...  \\\n",
       "\n",
       "   attributes.longitude attributes.biosample  attributes.latitude   \n",
       "0              -76.4716         SAMEA4981828              44.6033  \\\n",
       "\n",
       "  attributes.accession attributes.analysis-completed   \n",
       "0           ERS2799892                    2019-03-12  \\\n",
       "\n",
       "  attributes.collection-date attributes.geo-loc-name attributes.sample-desc   \n",
       "0                 2017-01-01                  Canada                 08-152  \\\n",
       "\n",
       "  attributes.environment-biome attributes.environment-feature   \n",
       "0             Mixedwood_Plains                  Envo:00000020  \\\n",
       "\n",
       "  attributes.environment-material attributes.sample-name   \n",
       "0                   Envo:00002011                 08-152  \\\n",
       "\n",
       "  attributes.sample-alias attributes.last-update   \n",
       "0                  08-152    2023-07-03T10:07:25  \\\n",
       "\n",
       "                  relationships.biome.data.id relationships.biome.data.type   \n",
       "0  root:Environmental:Aquatic:Freshwater:Lake                        biomes  \\\n",
       "\n",
       "                    relationships.studies.data  \n",
       "0  [{'id': 'MGYS00003941', 'type': 'studies'}]  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from jsonapi_client import Session\n",
    "import pandas as pd\n",
    "\n",
    "with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    samples = map(lambda r: r.json, mgnify.iterate(f'samples/ERS2799892'))\n",
    "    samples = pd.json_normalize(samples)\n",
    "\n",
    "samples"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4eaa0489",
   "metadata": {},
   "source": [
    "# Get files of complete study data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "49a802bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.alias</th>\n",
       "      <th>attributes.file-format.name</th>\n",
       "      <th>attributes.file-format.extension</th>\n",
       "      <th>attributes.file-format.compression</th>\n",
       "      <th>attributes.description.label</th>\n",
       "      <th>attributes.description.description</th>\n",
       "      <th>attributes.group-type</th>\n",
       "      <th>attributes.file-checksum.checksum</th>\n",
       "      <th>attributes.file-checksum.checksum-algorithm</th>\n",
       "      <th>relationships.pipeline.data.id</th>\n",
       "      <th>relationships.pipeline.data.type</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>study-downloads</td>\n",
       "      <td>ERP141089_GO-slim_abundances_v5.0.tsv</td>\n",
       "      <td>ERP141089_GO-slim_abundances_v5.0.tsv</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>False</td>\n",
       "      <td>GO slim annotation</td>\n",
       "      <td>GO slim annotation</td>\n",
       "      <td>Functional analysis</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.0</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/stud...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>study-downloads</td>\n",
       "      <td>ERP141089_GO_abundances_v5.0.tsv</td>\n",
       "      <td>ERP141089_GO_abundances_v5.0.tsv</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>False</td>\n",
       "      <td>Complete GO annotation</td>\n",
       "      <td>Complete GO annotation</td>\n",
       "      <td>Functional analysis</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.0</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/stud...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>study-downloads</td>\n",
       "      <td>ERP141089_IPR_abundances_v5.0.tsv</td>\n",
       "      <td>ERP141089_IPR_abundances_v5.0.tsv</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>False</td>\n",
       "      <td>InterPro matches</td>\n",
       "      <td>InterPro matches (TSV)</td>\n",
       "      <td>Functional analysis</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.0</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/stud...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>study-downloads</td>\n",
       "      <td>ERP141089_phylum_taxonomy_abundances_SSU_v5.0.tsv</td>\n",
       "      <td>ERP141089_phylum_taxonomy_abundances_SSU_v5.0.tsv</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>False</td>\n",
       "      <td>Phylum level taxonomies SSU</td>\n",
       "      <td>Phylum level taxonomies SSU (TSV)</td>\n",
       "      <td>Taxonomic analysis SSU rRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.0</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/stud...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>study-downloads</td>\n",
       "      <td>ERP141089_taxonomy_abundances_SSU_v5.0.tsv</td>\n",
       "      <td>ERP141089_taxonomy_abundances_SSU_v5.0.tsv</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>False</td>\n",
       "      <td>Taxonomic assignments SSU</td>\n",
       "      <td>Taxonomic assignments SSU (TSV)</td>\n",
       "      <td>Taxonomic analysis SSU rRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.0</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/stud...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>study-downloads</td>\n",
       "      <td>ERP141089_phylum_taxonomy_abundances_LSU_v5.0.tsv</td>\n",
       "      <td>ERP141089_phylum_taxonomy_abundances_LSU_v5.0.tsv</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>False</td>\n",
       "      <td>Phylum level taxonomies LSU</td>\n",
       "      <td>Phylum level taxonomies LSU (TSV)</td>\n",
       "      <td>Taxonomic analysis LSU rRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.0</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/stud...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>study-downloads</td>\n",
       "      <td>ERP141089_taxonomy_abundances_LSU_v5.0.tsv</td>\n",
       "      <td>ERP141089_taxonomy_abundances_LSU_v5.0.tsv</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>False</td>\n",
       "      <td>Taxonomic assignments LSU</td>\n",
       "      <td>Taxonomic assignments LSU (TSV)</td>\n",
       "      <td>Taxonomic analysis LSU rRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.0</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/stud...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              type                                                 id   \n",
       "0  study-downloads              ERP141089_GO-slim_abundances_v5.0.tsv  \\\n",
       "0  study-downloads                   ERP141089_GO_abundances_v5.0.tsv   \n",
       "0  study-downloads                  ERP141089_IPR_abundances_v5.0.tsv   \n",
       "0  study-downloads  ERP141089_phylum_taxonomy_abundances_SSU_v5.0.tsv   \n",
       "0  study-downloads         ERP141089_taxonomy_abundances_SSU_v5.0.tsv   \n",
       "0  study-downloads  ERP141089_phylum_taxonomy_abundances_LSU_v5.0.tsv   \n",
       "0  study-downloads         ERP141089_taxonomy_abundances_LSU_v5.0.tsv   \n",
       "\n",
       "                                    attributes.alias   \n",
       "0              ERP141089_GO-slim_abundances_v5.0.tsv  \\\n",
       "0                   ERP141089_GO_abundances_v5.0.tsv   \n",
       "0                  ERP141089_IPR_abundances_v5.0.tsv   \n",
       "0  ERP141089_phylum_taxonomy_abundances_SSU_v5.0.tsv   \n",
       "0         ERP141089_taxonomy_abundances_SSU_v5.0.tsv   \n",
       "0  ERP141089_phylum_taxonomy_abundances_LSU_v5.0.tsv   \n",
       "0         ERP141089_taxonomy_abundances_LSU_v5.0.tsv   \n",
       "\n",
       "  attributes.file-format.name attributes.file-format.extension   \n",
       "0                         TSV                              tsv  \\\n",
       "0                         TSV                              tsv   \n",
       "0                         TSV                              tsv   \n",
       "0                         TSV                              tsv   \n",
       "0                         TSV                              tsv   \n",
       "0                         TSV                              tsv   \n",
       "0                         TSV                              tsv   \n",
       "\n",
       "   attributes.file-format.compression attributes.description.label   \n",
       "0                               False           GO slim annotation  \\\n",
       "0                               False       Complete GO annotation   \n",
       "0                               False             InterPro matches   \n",
       "0                               False  Phylum level taxonomies SSU   \n",
       "0                               False    Taxonomic assignments SSU   \n",
       "0                               False  Phylum level taxonomies LSU   \n",
       "0                               False    Taxonomic assignments LSU   \n",
       "\n",
       "  attributes.description.description        attributes.group-type   \n",
       "0                 GO slim annotation          Functional analysis  \\\n",
       "0             Complete GO annotation          Functional analysis   \n",
       "0             InterPro matches (TSV)          Functional analysis   \n",
       "0  Phylum level taxonomies SSU (TSV)  Taxonomic analysis SSU rRNA   \n",
       "0    Taxonomic assignments SSU (TSV)  Taxonomic analysis SSU rRNA   \n",
       "0  Phylum level taxonomies LSU (TSV)  Taxonomic analysis LSU rRNA   \n",
       "0    Taxonomic assignments LSU (TSV)  Taxonomic analysis LSU rRNA   \n",
       "\n",
       "  attributes.file-checksum.checksum   \n",
       "0                                    \\\n",
       "0                                     \n",
       "0                                     \n",
       "0                                     \n",
       "0                                     \n",
       "0                                     \n",
       "0                                     \n",
       "\n",
       "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
       "0                                                                        5.0  \\\n",
       "0                                                                        5.0   \n",
       "0                                                                        5.0   \n",
       "0                                                                        5.0   \n",
       "0                                                                        5.0   \n",
       "0                                                                        5.0   \n",
       "0                                                                        5.0   \n",
       "\n",
       "  relationships.pipeline.data.type   \n",
       "0                        pipelines  \\\n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "\n",
       "                                                 url  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from jsonapi_client import Session\n",
    "import pandas as pd\n",
    "\n",
    "#pd.set_option('display.max_colwidth', None)\n",
    "\n",
    "with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "\n",
    "    dfs = []\n",
    "    for r in mgnify.iterate(f'studies/MGYS00006137/downloads'):\n",
    "        df = pd.json_normalize(r.json)\n",
    "        df['url'] = str(r.links.self)\n",
    "        dfs.append(df)\n",
    "\n",
    "main_df = pd.concat(dfs)\n",
    "main_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "6bfaab6c",
   "metadata": {},
   "source": [
    "# Put it together and download the analysis data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c4af44d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MGYS00005611\n",
      "              type                                                 id   \n",
      "0  study-downloads  SRP252220_phylum_taxonomy_abundances_SSU_v5.0.tsv  \\\n",
      "\n",
      "                                    attributes.alias   \n",
      "0  SRP252220_phylum_taxonomy_abundances_SSU_v5.0.tsv  \\\n",
      "\n",
      "  attributes.file-format.name attributes.file-format.extension   \n",
      "0                         TSV                              tsv  \\\n",
      "\n",
      "   attributes.file-format.compression attributes.description.label   \n",
      "0                               False  Phylum level taxonomies SSU  \\\n",
      "\n",
      "  attributes.description.description        attributes.group-type   \n",
      "0  Phylum level taxonomies SSU (TSV)  Taxonomic analysis SSU rRNA  \\\n",
      "\n",
      "  attributes.file-checksum.checksum   \n",
      "0                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        5.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
      "              type                                          id   \n",
      "0  study-downloads  SRP252220_taxonomy_abundances_SSU_v5.0.tsv  \\\n",
      "\n",
      "                             attributes.alias attributes.file-format.name   \n",
      "0  SRP252220_taxonomy_abundances_SSU_v5.0.tsv                         TSV  \\\n",
      "\n",
      "  attributes.file-format.extension  attributes.file-format.compression   \n",
      "0                              tsv                               False  \\\n",
      "\n",
      "  attributes.description.label attributes.description.description   \n",
      "0    Taxonomic assignments SSU    Taxonomic assignments SSU (TSV)  \\\n",
      "\n",
      "         attributes.group-type attributes.file-checksum.checksum   \n",
      "0  Taxonomic analysis SSU rRNA                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        5.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
      "              type                                                 id   \n",
      "0  study-downloads  SRP252220_phylum_taxonomy_abundances_LSU_v5.0.tsv  \\\n",
      "\n",
      "                                    attributes.alias   \n",
      "0  SRP252220_phylum_taxonomy_abundances_LSU_v5.0.tsv  \\\n",
      "\n",
      "  attributes.file-format.name attributes.file-format.extension   \n",
      "0                         TSV                              tsv  \\\n",
      "\n",
      "   attributes.file-format.compression attributes.description.label   \n",
      "0                               False  Phylum level taxonomies LSU  \\\n",
      "\n",
      "  attributes.description.description        attributes.group-type   \n",
      "0  Phylum level taxonomies LSU (TSV)  Taxonomic analysis LSU rRNA  \\\n",
      "\n",
      "  attributes.file-checksum.checksum   \n",
      "0                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        5.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
      "              type                                          id   \n",
      "0  study-downloads  SRP252220_taxonomy_abundances_LSU_v5.0.tsv  \\\n",
      "\n",
      "                             attributes.alias attributes.file-format.name   \n",
      "0  SRP252220_taxonomy_abundances_LSU_v5.0.tsv                         TSV  \\\n",
      "\n",
      "  attributes.file-format.extension  attributes.file-format.compression   \n",
      "0                              tsv                               False  \\\n",
      "\n",
      "  attributes.description.label attributes.description.description   \n",
      "0    Taxonomic assignments LSU    Taxonomic assignments LSU (TSV)  \\\n",
      "\n",
      "         attributes.group-type attributes.file-checksum.checksum   \n",
      "0  Taxonomic analysis LSU rRNA                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        5.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
      "MGYS00001499\n",
      "              type                                             id   \n",
      "0  study-downloads  ERP013549_phylum_taxonomy_abundances_v3.0.tsv  \\\n",
      "\n",
      "                                attributes.alias attributes.file-format.name   \n",
      "0  ERP013549_phylum_taxonomy_abundances_v3.0.tsv                         TSV  \\\n",
      "\n",
      "  attributes.file-format.extension  attributes.file-format.compression   \n",
      "0                              tsv                               False  \\\n",
      "\n",
      "  attributes.description.label attributes.description.description   \n",
      "0      Phylum level taxonomies      Phylum level taxonomies (TSV)  \\\n",
      "\n",
      "  attributes.group-type attributes.file-checksum.checksum   \n",
      "0    Taxonomic analysis                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        3.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
      "              type                                      id   \n",
      "0  study-downloads  ERP013549_taxonomy_abundances_v3.0.tsv  \\\n",
      "\n",
      "                         attributes.alias attributes.file-format.name   \n",
      "0  ERP013549_taxonomy_abundances_v3.0.tsv                         TSV  \\\n",
      "\n",
      "  attributes.file-format.extension  attributes.file-format.compression   \n",
      "0                              tsv                               False  \\\n",
      "\n",
      "  attributes.description.label attributes.description.description   \n",
      "0        Taxonomic assignments        Taxonomic assignments (TSV)  \\\n",
      "\n",
      "  attributes.group-type attributes.file-checksum.checksum   \n",
      "0    Taxonomic analysis                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        3.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
      "              type                            id   \n",
      "0  study-downloads  ERP013549_diversity_v3.0.tsv  \\\n",
      "\n",
      "               attributes.alias attributes.file-format.name   \n",
      "0  ERP013549_diversity_v3.0.tsv                         TSV  \\\n",
      "\n",
      "  attributes.file-format.extension  attributes.file-format.compression   \n",
      "0                              tsv                               False  \\\n",
      "\n",
      "  attributes.description.label attributes.description.description   \n",
      "0  Taxonomic diversity metrics  Taxonomic diversity metrics (TSV)  \\\n",
      "\n",
      "  attributes.group-type attributes.file-checksum.checksum   \n",
      "0            Statistics                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        3.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
      "              type                                                 id   \n",
      "0  study-downloads  ERP013549_phylum_taxonomy_abundances_SSU_v5.0.tsv  \\\n",
      "\n",
      "                                    attributes.alias   \n",
      "0  ERP013549_phylum_taxonomy_abundances_SSU_v5.0.tsv  \\\n",
      "\n",
      "  attributes.file-format.name attributes.file-format.extension   \n",
      "0                         TSV                              tsv  \\\n",
      "\n",
      "   attributes.file-format.compression attributes.description.label   \n",
      "0                               False  Phylum level taxonomies SSU  \\\n",
      "\n",
      "  attributes.description.description        attributes.group-type   \n",
      "0  Phylum level taxonomies SSU (TSV)  Taxonomic analysis SSU rRNA  \\\n",
      "\n",
      "  attributes.file-checksum.checksum   \n",
      "0                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        5.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
      "              type                                          id   \n",
      "0  study-downloads  ERP013549_taxonomy_abundances_SSU_v5.0.tsv  \\\n",
      "\n",
      "                             attributes.alias attributes.file-format.name   \n",
      "0  ERP013549_taxonomy_abundances_SSU_v5.0.tsv                         TSV  \\\n",
      "\n",
      "  attributes.file-format.extension  attributes.file-format.compression   \n",
      "0                              tsv                               False  \\\n",
      "\n",
      "  attributes.description.label attributes.description.description   \n",
      "0    Taxonomic assignments SSU    Taxonomic assignments SSU (TSV)  \\\n",
      "\n",
      "         attributes.group-type attributes.file-checksum.checksum   \n",
      "0  Taxonomic analysis SSU rRNA                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        5.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
      "              type                                                 id   \n",
      "0  study-downloads  ERP013549_phylum_taxonomy_abundances_itsonedb_...  \\\n",
      "\n",
      "                                    attributes.alias   \n",
      "0  ERP013549_phylum_taxonomy_abundances_itsonedb_...  \\\n",
      "\n",
      "  attributes.file-format.name attributes.file-format.extension   \n",
      "0                         TSV                              tsv  \\\n",
      "\n",
      "   attributes.file-format.compression      attributes.description.label   \n",
      "0                               False  Phylum level taxonomies ITSoneDB  \\\n",
      "\n",
      "       attributes.description.description        attributes.group-type   \n",
      "0  Phylum level taxonomies ITSoneDB (TSV)  Taxonomic analysis ITSoneDB  \\\n",
      "\n",
      "  attributes.file-checksum.checksum   \n",
      "0                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        5.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
      "              type                                               id   \n",
      "0  study-downloads  ERP013549_taxonomy_abundances_itsonedb_v5.0.tsv  \\\n",
      "\n",
      "                                  attributes.alias   \n",
      "0  ERP013549_taxonomy_abundances_itsonedb_v5.0.tsv  \\\n",
      "\n",
      "  attributes.file-format.name attributes.file-format.extension   \n",
      "0                         TSV                              tsv  \\\n",
      "\n",
      "   attributes.file-format.compression    attributes.description.label   \n",
      "0                               False  Taxonomic assignments ITSoneDB  \\\n",
      "\n",
      "     attributes.description.description        attributes.group-type   \n",
      "0  Taxonomic assignments ITSoneDB (TSV)  Taxonomic analysis ITSoneDB  \\\n",
      "\n",
      "  attributes.file-checksum.checksum   \n",
      "0                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        5.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
      "              type                                                 id   \n",
      "0  study-downloads  ERP013549_phylum_taxonomy_abundances_unite_v5....  \\\n",
      "\n",
      "                                    attributes.alias   \n",
      "0  ERP013549_phylum_taxonomy_abundances_unite_v5....  \\\n",
      "\n",
      "  attributes.file-format.name attributes.file-format.extension   \n",
      "0                         TSV                              tsv  \\\n",
      "\n",
      "   attributes.file-format.compression   attributes.description.label   \n",
      "0                               False  Phylum level taxonomies UNITE  \\\n",
      "\n",
      "    attributes.description.description     attributes.group-type   \n",
      "0  Phylum level taxonomies UNITE (TSV)  Taxonomic analysis UNITE  \\\n",
      "\n",
      "  attributes.file-checksum.checksum   \n",
      "0                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        5.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n",
      "              type                                            id   \n",
      "0  study-downloads  ERP013549_taxonomy_abundances_unite_v5.0.tsv  \\\n",
      "\n",
      "                               attributes.alias attributes.file-format.name   \n",
      "0  ERP013549_taxonomy_abundances_unite_v5.0.tsv                         TSV  \\\n",
      "\n",
      "  attributes.file-format.extension  attributes.file-format.compression   \n",
      "0                              tsv                               False  \\\n",
      "\n",
      "  attributes.description.label attributes.description.description   \n",
      "0  Taxonomic assignments UNITE  Taxonomic assignments UNITE (TSV)  \\\n",
      "\n",
      "      attributes.group-type attributes.file-checksum.checksum   \n",
      "0  Taxonomic analysis UNITE                                    \\\n",
      "\n",
      "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id   \n",
      "0                                                                        5.0  \\\n",
      "\n",
      "  relationships.pipeline.data.type   \n",
      "0                        pipelines  \\\n",
      "\n",
      "                                                 url  \n",
      "0  https://www.ebi.ac.uk/metagenomics/api/v1/stud...  \n"
     ]
    }
   ],
   "source": [
    "from jsonapi_client import Session\n",
    "import pandas as pd\n",
    "import requests\n",
    "import os\n",
    "\n",
    "study_accessions = [\"MGYS00005611\", \"MGYS00001499\"]\n",
    "data_type = \"Taxonomic assignments SSU\"\n",
    "\n",
    "data_output_folder = 'data/SSU'\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "for study_accession in study_accessions:\n",
    "\n",
    "    print(study_accession)\n",
    "\n",
    "    # get df of metadata and urls to data \n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "\n",
    "        dfs = []\n",
    "        for r in mgnify.iterate(f'studies/{study_accession}/downloads'):\n",
    "            df = pd.json_normalize(r.json)\n",
    "            df['url'] = str(r.links.self)\n",
    "            dfs.append(df)\n",
    "\n",
    "    # df of all data that can be downloaded for this study\n",
    "    main_df = pd.concat(dfs)\n",
    "\n",
    "    # get specific data table\n",
    "    url = main_df.loc[main_df[\"attributes.description.label\"] == data_type, \"url\"].iloc[0]\n",
    "    response = requests.get(url)\n",
    "\n",
    "    data_output_path = os.path.join(output_folder, f\"{study_accession}.txt\")\n",
    "    with open(data_output_path, \"w\") as f:\n",
    "        f.write(response.text)\n",
    "\n",
    "    # get analysis metadata\n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "        analyses = map(lambda r: r.json, mgnify.iterate(f'studies/{study_accession}/analyses'))\n",
    "        analyses = pd.json_normalize(analyses)\n",
    "\n",
    "    meta_data_output_path = os.path.join(output_folder, f\"{study_accession}.metadata.txt\")\n",
    "    analyses.to_csv(meta_data_output_path)\n",
    "\n",
    "    # TODO add metadata of each sample (geolocation, biome)\n",
    "    # Ideally also the ENA accession (phenotype information)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "45d0d4f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(558, 117)\n",
      "(57, 26)\n",
      "(589, 143)\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# merge dfs to have identical rows\n",
    "\n",
    "def create_identical_index(output_folder, mode = \"outer\"):\n",
    "    \"\"\"\n",
    "    Converts csv files with different indexes to files with identical and sorted indexes, using \n",
    "    either inner or outer join merge based on the index\n",
    "    \"\"\"\n",
    "\n",
    "    dfs = []\n",
    "    columns_per_file = {}\n",
    "    for file in os.listdir(output_folder):\n",
    "        file_path = os.path.join(output_folder,file)\n",
    "        df = pd.read_csv(file_path, sep=\"\\t\", index_col=0)\n",
    "        # df.set_index(0, inplace=True)\n",
    "        df.index.name = \"features\"\n",
    "        columns_per_file[file] = list(df.columns)\n",
    "        dfs.append(df)\n",
    "        print(df.shape)\n",
    "\n",
    "    if mode == \"outer\":\n",
    "        #features that do not exist in one of the dfs get filled with zero\n",
    "        main_df = pd.concat(dfs, join=mode, axis = 1)\n",
    "        main_df.fillna(0, inplace=True)\n",
    "        print(main_df.shape)\n",
    "\n",
    "    elif mode == \"inner\":\n",
    "        #only features that exist in both dfs are kept\n",
    "        main_df = pd.concat(dfs, join=mode, axis = 1)\n",
    "        print(main_df.shape)\n",
    "\n",
    "    # print(main_df.index.is_unique)\n",
    "\n",
    "    new_output = output_folder + '_mod'\n",
    "    os.makedirs(new_output, exist_ok=True)\n",
    "\n",
    "    #write back to file\n",
    "    for file, columns in columns_per_file.items():\n",
    "        df = main_df.loc[:,columns]\n",
    "        # file_name, file_ending = file.split(\".\")[0], file.split(\".\")[1]\n",
    "        out_file = os.path.join(new_output, file)\n",
    "        df.to_csv(out_file)\n",
    "\n",
    "create_identical_index(output_folder, mode=\"outer\")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "bfb83410",
   "metadata": {},
   "source": [
    "# Get phylum level study data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71fc2c51",
   "metadata": {},
   "outputs": [],
   "source": [
    "from jsonapi_client import Session\n",
    "import pandas as pd\n",
    "import requests\n",
    "import os\n",
    "\n",
    "study_accessions = [\"MGYS00005591\",\"MGYS00005745\",\"MGYS00005745\",\"MGYS00005601\",\"MGYS00005755\", \"MGYS00005755\", \"MGYS00005742\"]\n",
    "#data_type = \"Taxonomic assignments SSU\"\n",
    "data_type = \"Phylum level taxonomies LSU\"\n",
    "\n",
    "\n",
    "data_output_folder = 'data/SSU'\n",
    "os.makedirs(data_output_folder, exist_ok=True)\n",
    "\n",
    "for study_accession in study_accessions:\n",
    "\n",
    "    print(study_accession)\n",
    "\n",
    "    # get df of metadata and urls to data \n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "\n",
    "        dfs = []\n",
    "        for r in mgnify.iterate(f'studies/{study_accession}/downloads'):\n",
    "            df = pd.json_normalize(r.json)\n",
    "            df['url'] = str(r.links.self)\n",
    "            dfs.append(df)\n",
    "\n",
    "    # df of all data that can be downloaded for this study\n",
    "    main_df = pd.concat(dfs)\n",
    "\n",
    "    # get specific data table\n",
    "    try:\n",
    "        url = main_df.loc[main_df[\"attributes.description.label\"] == data_type, \"url\"].iloc[0]\n",
    "        response = requests.get(url)\n",
    "\n",
    "        data_output_path = os.path.join(data_output_folder, f\"{study_accession}.txt\")\n",
    "        with open(data_output_path, \"w\") as f:\n",
    "            f.write(response.text)\n",
    "\n",
    "        # add to galaxy\n",
    "        put(data_output_path)\n",
    "    \n",
    "    except:\n",
    "        print(f\"Could not fetch data for: {study_accession} using the lable: {data_type}\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "213b5d7c",
   "metadata": {},
   "source": [
    "download_link = \"https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00006137/pipelines/5.0/file/ERP141089_GO-slim_abundances_v5.0.tsv\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
